raft/raft.go

```diff
@@ -31,9 +31,18 @@ import (
 	"go.etcd.io/etcd/raft/v3/tracker"
 )
 
-// None is a placeholder node ID used when there is no leader.
-const None uint64 = 0
-const noLimit = math.MaxUint64
+const (
+	// None is a placeholder node ID used when there is no leader.
+	None uint64 = 0
+	// LocalAppendThread is a reference to a local thread that saves unstable
+	// log entries and snapshots to stable storage. The identifier is used as a
+	// target for MsgStorageAppend messages when AsyncStorageWrites is enabled.
+	LocalAppendThread uint64 = math.MaxUint64
```

Comment:
There is a 1:1 relationship between the message type and the thread pseudo-node, but I have come to like the explicitness of the thread targets. They accurately represent the meaning of the messages, permit routing solely based on the `Message.To` field, and they make it easier to talk about ordering requirements to a given target without ordering requirements across targets.

At the same time, your concern about backwards compat is valid. We could get around the issue by only preventing the use of these IDs when AsyncStorageWrites is enabled.

Reply:
I agree that we don't strictly need the pseudo-targets and that clients using this library could get away without them as long as they carefully read documentation of the library which describes which message types flow in which directions and how to route based on message types.

In my opinion, the pseudo-targets are a form of documentation so that we don't need to push those concerns onto users. They also make the message-passing interface consistent. Without them, users of the library are forced to ask:
* what does it mean for a message to have no target? Who do I send it to? Why do some messages not have targets?
* is `Msg.To: r.ID` a reference to the local `etcd/raft` state machine library or a reference to a local storage component. How do I distinguish the two?
* if we're not even assigning `Msg.To: r.ID` on `Response` messages, am I also matching on `MsgStorageAppendResp` and `MsgStorageApplyResp`? If so, why do `Response` slices contain some ID-ed messages and some not? Why do I sometimes (on a leader) have `Responses: [{MsgAppResp, To: r.ID}, {MsgStorageAppendResp, To: 0}]`?
* which local messages can be reordered and which need to be acted on in order?

The question to ask here isn't "do we need these IDs?", it's "how much harder would this be to explain without them?"

Reply:
I could write a lengthy response here making the opposite case, but I think both arguments have merit and the choice is ultimately subjective. Let's go with your preference.